### PR TITLE
Create A Diagnostic Callback interface for Metricsets

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This PR adds a special interface, called `DiagnosticSet`, to metricbeat, that allows individual metricsets to report data that can be gathered by `elastic-agent diagnostics`. This involves minimal change to the runners and reloaders, but is otherwise fairly simple. Any metricset that wants to implement this interface merely needs to add a `DiagnosticSetup` method to its MetricSet object:

```
func (m *MetricSet) DiagnosticSetup() []diagnostics.DiagnosticSetup {
	m.Logger().Infof("got DiagnosticSetup request for system/core")
	if runtime.GOOS == "linux" {
		return []diagnostics.DiagnosticSetup{{
			Name:        "core-stat",
			Description: "/proc/stat file",
			Filename:    "stat",
			Callback:    m.getDiagData,
		}}
	} else {
		return nil
	}

}
```

In addition to adding a new interface, this PR also implements `DiagnosticSetup` for a handful of system modules where additional debug data often comes in handy. This also removes the `module.Runner` interface in favor of `cfgfile.Runner`, as the two interfaces are apparently identical.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- Pull down this PR
- Build elastic-agent with `EXTERNAL=false mage dev:package` to build in this beat version
- In run elastic-agent, however you normally would, then gather diagnostics via `elastic-agent diagnostics`

### CODE REVIEW

File: libbeat/common/diagnostics/diagnostics.go
Comment (faec) [inline libbeat/common/diagnostics/diagnostics.go:38]: Is `Name` used for anything? It seems like we might just need `Filename`

File: libbeat/common/diagnostics/diagnostics.go
Comment (faec) [inline libbeat/common/diagnostics/diagnostics.go:47]: My big-picture wish is that we could skip the whole `Callback` indirection and just return a list of filenames + data directly from the main `Diagnostics()` call -- that is, have the call return the actual current diagnostics instead of a list of possible future diagnostics to add hooks for. I don't know of a reason to register every possible file in advance when we could just return a list of the relevant ones when `diagnostics` is actually invoked, since the registration seems to only be used internally to the client, not by Agent itself. But it would require new API in `elastic-agent-client`, so this approach is probably necessary until we address that...

File: libbeat/common/diagnostics/diagnostics.go
Comment (fearful-symmetry) [reply libbeat/common/diagnostics/diagnostics.go:28]: Yah, when I started out they actually _were_ separate interfaces, then they weren't. Now that I think of it, probably worth changing, yah.

File: metricbeat/mb/module/runner_test.go
Comment (fearful-symmetry) [reply metricbeat/mb/module/runner_test.go:99]: Not sure what you mean? `ModuleDiagnostics` should return an empty list on non-linux systems in that case.

File: libbeat/cfgfile/list.go
Comment (fearful-symmetry) [reply libbeat/cfgfile/list.go:105]: So, you can do `errors.Is(err, &common.ErrInputNotFinished{})`, but I don't know if that'll produce the same result as a typecast, since I assume `errors.Is` is comparing the values, and not just the types?

File: libbeat/common/diagnostics/diagnostics.go
Comment (fearful-symmetry) [reply libbeat/common/diagnostics/diagnostics.go:38]: So, it's used as a field in the client `RegisterDiagnosticHook()`, and I wanted this type to be a somewhat un-opinionated wrapper for that.

File: libbeat/common/diagnostics/diagnostics.go
Comment (fearful-symmetry) [reply libbeat/common/diagnostics/diagnostics.go:47]: Yup, agreed. Would be simpler to just not do the callback thing, but I didn't want to invent some kind of overly-clever wrapper around `RegisterDiagnosticHook`.

File: metricbeat/mb/module/runner_test.go
Comment (faec) [reply metricbeat/mb/module/runner_test.go:100]: Well, the part that confuses me is that `require.NotEmpty(t, diags)` is outside the `runtime.GOOS == "linux"` check, which seems to suggest it should be nonempty on all platforms

File: libbeat/cfgfile/list.go
Comment (faec) [reply libbeat/cfgfile/list.go:107]: (long sigh)... well, the nice thing to do then would be to implement `Is` on `ErrInputNotFinished` since value-equality is probably not useful for this type of error (see https://pkg.go.dev/errors#Is). But that is arguably out of scope here so I'm ok with the lint workaround for now 😜

File: metricbeat/mb/module/runner_test.go
Comment (faec) [reply metricbeat/mb/module/runner_test.go:100]: I guess what I'd like to see on non-linux is either a `require.Empty(t, diags)` or, if the semantics are such that it always returns the initial callback (which seems like it based on this code?), a `require.Empty(t, diags[0].Callback())`

**Review by faec — APPROVED** [review]
Looks good, approved pending the remaining testing nit

File: metricbeat/mb/module/runner_test.go
Comment (fearful-symmetry) [reply metricbeat/mb/module/runner_test.go:100]: Ah, okay! Yah, agreed, should add a `require.Empty` section. That whole test is a tad confusing, should probably document it a bit better.

File: libbeat/cfgfile/list.go
Comment (fearful-symmetry) [reply libbeat/cfgfile/list.go:107]: Yah, it's definitely kinda awkward, I was just really hesitant to tinker around with some unrelated code that I've never touched, since that's bitten me before.

**Comment by pierrehilbert** [comment]
@fearful-symmetry are we still waiting for something before merging this one?

**Comment by fearful-symmetry** [comment]
@pierrehilbert nope, just got distracted yesterday.

**Pull request merged**
